libraries/AP_Landing/AP_Landing_Deepstall.cpp

```diff
@@ -0,0 +1,380 @@
+
+
+    GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "Predict: %f %f", stall_distance,  estimated_forward * height / type_deepstall_down_speed + stall_distance);
+
+    return estimated_forward * height / type_deepstall_down_speed + stall_distance;
+}
+
+bool AP_Landing::type_deepstall_verify_breakout(const Location &current_loc, const Location &target_loc,
+                                               const float height_error) const
+{
+    Vector2f location_delta = location_diff(current_loc, target_loc);
+    float heading_error = degrees(ahrs.groundspeed_vector().angle(location_delta));
+
+    // Check to see if the the plane is heading toward the land waypoint. We use 20 degrees (+/-10 deg)
+    // of margin so that the altitude to be within 5 meters of desired
+
+    if (heading_error <= 10.0  && fabsf(height_error) < DEEPSTALL_LOITER_ALT_TOLERANCE) {
+            // Want to head in a straight line from _here_ to the next waypoint instead of center of loiter wp
+            return true;
+    }   
+    return false;
+}
+
+float AP_Landing::type_deepstall_update_steering()
+{
+    Location current_loc;
+    if (!ahrs.get_position(current_loc)) {
+        // panic if no position source is available
+        // continue the deepstall. but target just holding the wings held level as deepstall should be a minimal energy
+        // configuration on the aircraft, and if a position isn't available aborting would be worse
+        GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_CRITICAL, "Deepstall: No position available. Attempting to hold level");
+        memcpy(&current_loc, &type_deepstall_landing_point, sizeof(Location));
+    }
+    uint32_t time = AP_HAL::millis();
+    float dt = constrain_float(time - type_deepstall_last_time, (uint32_t)10UL, (uint32_t)200UL) / 1000.0;
+    type_deepstall_last_time = time;
+
+
+    Vector2f ab = location_diff(type_deepstall_arc_exit, type_deepstall_extended_approach);
+    ab.normalize();
+    Vector2f a_air = location_diff(type_deepstall_arc_exit, current_loc);
+
+    float crosstrack_error = a_air % ab;
+    float sine_nu1 = constrain_float(crosstrack_error / MAX(type_deepstall_l1_period, 0.1f), -0.7071f, 0.7107f);
+    float nu1 = asinf(sine_nu1);
+
+    if (type_deepstall_l1_i > 0) {
+        type_deepstall_l1_xtrack_i += nu1 * type_deepstall_l1_i / dt;
+        type_deepstall_l1_xtrack_i = constrain_float(type_deepstall_l1_xtrack_i, -0.5f, 0.5f);
+        nu1 += type_deepstall_l1_xtrack_i;
+    }
+
+    float desired_change = wrap_PI(radians(type_deepstall_target_heading_deg) + nu1 - ahrs.yaw);
+
+    float yaw_rate = ahrs.get_gyro().z;
+    float yaw_rate_limit = radians(type_deepstall_yaw_rate_limit);
+    float error = wrap_PI(constrain_float(desired_change / type_deepstall_time_constant,
+                                          -yaw_rate_limit, yaw_rate_limit) - yaw_rate);
+
+    GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "x: %f e: %f r: %f d: %f",
```

Comment:
more debug? This sort of thing should have #if 0 around it because it's clearly only for developers. May want to consider that for most of the other GCS strings too. This library is very verbose.